### ISSUE TITLE
False positive DocblockTypeContradiction for specific class

### ISSUE THREAD

**Issue by bugreportuser** [vimeo/psalm#1338]
Psalm expects the below example to print nothing, but it prints `func2`. Surprisingly, the example is minimal.

Example (https://getpsalm.org/r/262b650f6b):
```php
<?php
final class Clazz {
    /**
     * @var bool
     */
    public $x;

    /**
     * @var int
     */
    public $y = 0;

    public function func1 (): bool {
        if ($this->y || rand(0, 1) !== 2) {
            return true;
        }
        return false;
    }

    public function func2 (): int {
        if ($this->y) {
            echo 'func2';
            return 1;
        }
        return 2;
    }

    public function __construct () {
        $this->x = false;
        if ($this->func1()) {
            $this->y = $this->func2();
        }
        $this->func2();
    }
}

new Clazz();
```

Example output:
```
Psalm output (using commit 356a45a):

ERROR: DocblockTypeContradiction - 21:13 - Found a contradiction with a docblock-defined type when evaluating $this->y and trying to reconcile type 'int(0)' to !falsy
```

**Comment by muglug** [comment]
Thanks!

**Comment by bugreportuser** [comment]
Thank you too! That was fast.
